internal/gitserver/addrs.go

```diff
@@ -66,3 +78,75 @@ func addrForKey(key string, addrs []string) string {
 	serverIndex := binary.BigEndian.Uint64(sum[:]) % uint64(len(addrs))
 	return addrs[serverIndex]
 }
+
+type GitserverConns struct {
+	GitserverAddresses
+	// invariant: there is one conn for every
+	grpcConns map[string]connAndErr
+}
+
+func (g *GitserverConns) ConnForRepo(userAgent string, repo api.RepoName) (*grpc.ClientConn, error) {
+	addr := g.AddrForRepo(userAgent, repo)
+	ce := g.grpcConns[addr]
```

Comment:
Might be overly defensive, but maybe we protect against an invariant violation if grpcConns doesn't have an entry for addr: 

```suggestion
	ce, ok := g.grpcConns[addr]
	if !ok {
		return nil, errors.Newf("no gRPC connection found for address %q", addr) 
	}
```

Reply:
Oh, yeah, we've already got an error return, so no reason to just panic